R/tostring.R

```diff
@@ -179,6 +179,122 @@ do_cell_fnotes_wrap <- function(mat, widths, max_width, tf_wrap) {
+decimal_align <- function(string_mat, align_mat) {
+  # Evaluate if any values are to be decimal aligned
+  if (!any(grepl("dec", align_mat))) {
+    string_mat <- string_mat
+  } else {
+    for (i in seq(1, ncol(string_mat))) {
+      # Take a column and its decimal alignments
+      col_i <- as.character(string_mat[, i])
+      align_col_i <- grepl("dec", align_mat[, i])
+      # If no values are to be decimal aligned in the column (according
+      #  to the aligns matrix), or there are no numerical values, strings
+      #  remain as is
+      if (sum(align_col_i) == 0 ||
+          all(grepl("^[0-9]\\.", col_i))) {
+        string_mat[, i] <- string_mat[, i]
+      }
+
+      # Values to be decimal aligned
+      if (any(align_col_i)) {
+        # Extract values not to be aligned (NAs, non-numbers,
+        #  doesn't say "decimal" in alignment matrix)
+        nas <- vapply(col_i, is.na, FUN.VALUE = logical(1))
+        nonnum <- !grepl("[0-9]", col_i)
+        # Why not? grepl("[a-zA-Z]", col_i) # This excludes N=xx, e.g.
+
+        nonalign <- nas | nonnum | !align_col_i
+
+        # Do decimal alignment
+        if (length(col_i[!nonalign]) > 0) {
+
+          # Count the number of numbers in the string
+          matches <- gregexpr("\\d+\\.\\d+|\\d+", col_i[!nonalign])
+          more_than_one <- vapply(matches, function(x) {
+            sum(attr(x, "match.length") > 0) > 1
+          }, logical(1))
+
+          # Throw error in case any have more than 1 numbers
+          if (any(more_than_one)) {
+            stop("Decimal alignment is not supported for multiple values. ",
+                 "Found the following string with multiple numbers ",
+                 "(first 3 selected from column ", col_i[1],"): '",
+                 paste0(col_i[!nonalign][more_than_one][seq(1, 3)],
+                       collapse = "', '"), "'")
+          }
+
+          # General split (only one match -> the first)
+          main_regexp <- regexpr("\\d+", col_i[!nonalign])
+          left <- regmatches(col_i[!nonalign], main_regexp, invert = FALSE)
+          right <- regmatches(col_i[!nonalign], main_regexp, invert = TRUE)
+          right <- sapply(right, "[[", 2)
+
+          something_left <- sapply(strsplit(col_i[!nonalign], "\\d+"), "[[", 1)
+          left <- paste0(something_left, left)
+
+          if (!checkmate::test_set_equal(paste0(left, right), col_i[!nonalign]))
+            stop("Split string list lost some piece along the way. This ",
+                 "should not have happened. Please contact the maintainer.") # nocov
+
+          separator <- sapply(right, function(x) {
+            if (nzchar(x)) {
+              substr(x, 1, 1)
+            } else {
+              c(" ")
+            }
+          }, USE.NAMES = FALSE)
+
+          right <- sapply(right, function(x) {
+            if (nchar(x) > 1) {
+              substr(x, 2, nchar(x))
+            } else {
+              c("")
+            }
+          }, USE.NAMES = FALSE)
```

Comment:
This is the main change in algorithm. Now, it works for any supported formats (and I think beyond) @ayogasekaram